File: ProjectEuler/03_largest_prime_factor.c
/* What is the largest prime factor of 600851475143 */
/* Taken from https://projecteuler.net/problem=3 */

#include <stdio.h>

int main(void) {
    long number = 600851475143;
    long largest_prime_factor = 0;
    for (size_t i = 2; i <= number; i++) {
        if (number % i == 0) {
            if (i > largest_prime_factor) {
                largest_prime_factor = i;
            }

            number = number / i;
            i = 2;
        }
    }

    printf("Largest: %ld\n", largest_prime_factor);

    return 0;
}

File: ProjectEuler/03_largest_prime_factor.scala
/* What is the largest prime factor of 600851475143 */
/* Taken from https://projecteuler.net/problem=3 */

object LargestPrimeFactor {
  @annotation.tailrec
  def largest_prime_factor(quotient: BigInt, number: BigInt, largest: BigInt): BigInt = {
    if (quotient == 1 ||  number > quotient) {
      largest
    } else if (quotient % number == 0) {
      if (number > largest) {
        largest_prime_factor(quotient / number, 2, number)
      } else {
        largest_prime_factor(quotient / number, 2, largest)
      }
    } else {
      largest_prime_factor(quotient, number + 1, largest)
    }
  }

  def main(args: Array[String]): Unit = {
    var number = BigInt("600851475143")
    //    var result = largest_prime_factor(number, 2, 0);
    var result = largest_prime_factor(123123123, 2, 0);
    println(s"Resultado: $result")
  }
}
